lib/cc/workspace/path_tree.rb

```diff
@@ -0,0 +1,115 @@
+require "pathname"
+require "set"
+
+module CC
+  class Workspace
+    class PathTree
+      def initialize(root_path, children = {})
+        @root_path = root_path.dup.freeze
+        @children = children
+      end
+
+      def clone
+        self.class.new(root_path, children.dup)
+      end
+
+      def exclude_paths(paths)
+        paths.each { |path| delete(normalized_path_pieces(path)) }
+      end
+
+      def include_paths(paths)
+        paths.each { |path| insert(normalized_path_pieces(path)) }
+      end
+
+      def all_paths
+        if populated?
+          children.values.flat_map(&:all_paths)
+        else
+          [File.join(root_path, File::SEPARATOR)]
+        end
+      end
+
+      protected
+
+      def populated?
+        children.present?
+      end
+
+      def delete(path_pieces)
+        populate_all
+
+        if 1 == path_pieces.size
+          children.delete(path_pieces[0])
+        else
+          child = children[path_pieces[0]]
+          if child
+            child.delete(path_pieces.drop(1))
+            children.delete(path_pieces[0]) unless child.populated?
+          else
+            CLI.debug(
+              "Couldn't exclude because part of path doesn't exist or was already excluded.",
```

Comment:
Also for the cases where you might give the style of paths that `PathValidator` used to cover -- if you run something like `analyze /root/path` or `analyze ../../not/in/the/project/repo`, this will provide feedback about why that didn't work (albeit only with debug visibility).